Detectors/TOF/compression/src/CompressorTask.cxx

```diff
@@ -0,0 +1,88 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// @file   CompressorTask.cxx
+/// @author Roberto Preghenella
+/// @since  2019-12-18
+/// @brief  TOF raw data compressor task
+
+#include "TOFCompression/CompressorTask.h"
+#include "Framework/ControlService.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/WorkflowSpec.h"
+
+using namespace o2::framework;
+
+namespace o2
+{
+namespace tof
+{
+
+void CompressorTask::init(InitContext& ic)
+{
+  LOG(INFO) << "Compressor init";
+
+  /** link encoder output buffer **/
+  mCompressor.setEncoderBuffer(const_cast<char*>(mDataFrame.mBuffer));
+
+  auto decoderVerbose = ic.options().get<bool>("tof-compressor-decoder-verbose");
+  auto encoderVerbose = ic.options().get<bool>("tof-compressor-encoder-verbose");
+  auto checkerVerbose = ic.options().get<bool>("tof-compressor-checker-verbose");
+
+  mCompressor.setDecoderVerbose(decoderVerbose);
+  mCompressor.setEncoderVerbose(encoderVerbose);
+  mCompressor.setCheckerVerbose(checkerVerbose);
+
+  auto finishFunction = [this]() {
+    mCompressor.checkSummary();
+  };
+
+  ic.services().get<CallbackService>().set(CallbackService::Id::Stop, finishFunction);
+}
+
+void CompressorTask::run(ProcessingContext& pc)
+{
+  LOG(DEBUG) << "Compressor run";
+
+  /** receive input **/
+  for (auto& input : pc.inputs()) {
+    const auto* header = DataRefUtils::getHeader<o2::header::DataHeader*>(input);
+    auto payload = const_cast<char*>(input.payload);
+    auto payloadSize = header->payloadSize;
+    mCompressor.setDecoderBuffer(payload);
+    mCompressor.setDecoderBufferSize(payloadSize);
+
+    /** run **/
+    mCompressor.run();
+
+    /** push output **/
+    mDataFrame.mSize = mCompressor.getEncoderByteCounter();
+    pc.outputs().snapshot(Output{"TOF", "CMPDATAFRAME", 0, Lifetime::Timeframe}, mDataFrame);
```

Comment:
Hi Ruben,

I'm not sure I fully understand this, as I'm not really expert on DPL.
Notice on the other hand that this part was a bit of a home-made raw-data passing way for testing purposes.

I believe it should be perhaps made in a similar fashion as the `RawFileReader` passes data to DPL devices. I will have a look into that as well.

Cheers,
Roberto

Reply:
I asked @ktf, you basically need to declare your output with "wildcarded subspecification" (in your case subspecification is e.g. link ID) as 
````
Outputs{ OutputSpec(o2f::ConcreteDataTypeMatcher{"TOF", "CMPDATAFRAME"}) },
````
instead of https://github.com/AliceO2Group/AliceO2/blob/cdb07d7fdf8c16abc139aa18f58d47c8ad5ca3c2/Detectors/TOF/compression/src/CompressorTask.cxx#L79

and snapshot it as 
````
pc.outputs().snapshot(Output{"TOF", "CMPDATAFRAME", header->subSpecification, Lifetime::Timeframe}, mDataFrame);
````
instead of https://github.com/AliceO2Group/AliceO2/blob/cdb07d7fdf8c16abc139aa18f58d47c8ad5ca3c2/Detectors/TOF/compression/src/CompressorTask.cxx#L67

This is in theory and assuming that your ``mDataFrame`` corresponds to the data from a single link.

Reply:
Thanks Ruben,

I have completely ditched my `RawDataFrame` and followed the approach you used in the `RawFileReader` for sending the data
```
device->Send(parts, fairMQchannel);
```
which seems to be working ok.
I was not happy with using my class, this device should eventually send data to the `StfSender`  and to the `QC` device we will run on the FLP.

Cheers,
Roberto